common/src/main/scala/uk/ac/wellcome/models/Work.scala

```diff
@@ -15,7 +15,10 @@ trait Work extends Versioned {
   val lettering: Option[String]
   val createdDate: Option[Period]
   val subjects: List[AbstractConcept]
-  val creators: List[IdentityState[AbstractAgent]]
+
+  // Yes, this is kind of rubbish.  I haven't found another way to satisfy the compiler!
+  val contributors: List[Any]
```

Comment:
I’ve tried using `IdentityState[AbstractAgent]`, but the compiler doesn’t like it:

```
Work.scala:44: overriding value contributors in trait Work of type List[AbstractAgent]];
 value contributors has incompatible type
  contributors: List[Contributor[MaybeDisplayable[AbstractAgent]]] = Nil,
  ^
Work.scala:69: overriding value contributors in trait Work of type List[IdentityState[AbstractAgent]];
 value contributors has incompatible type
                          contributors: List[Contributor[Displayable[AbstractAgent]]] = Nil,
                          ^
two errors found
```

Reply:
Yeah, I don't think this works

I think it should be `List[Contributor[IdentityState[Agent]]]` here, `List[Contributor[MaybeDisplayable[Agent]]]` in `UnidentifiedWork` and `List[Contributor[Displayable[Agent]]]` in IdentifiedWork`

Reply:
Agent or AbstractAgent?

Reply:
AbstractAgent sorry. What error do you get doing `contributors: List[Contributor[IdentityState[AbstractAgent]]]` on `Work`

Reply:
The error I quoted above. I wasn’t very clear; it’s the error I get if I write:

```scala
val contributors: List[Contributor[IdentityState[AbstractAgent]]]
```

Reply:
From that errors it seems that contributors was set to `List[AbstractAgent]`... or at least it had a `List[AbtractAgent]` somewhere which is definitely wrong

Reply:
Okay, let me have another try.

Reply:
okay, so with this in `Work.scala`:

```scala
  // Yes, this is kind of rubbish.  I haven't found another way to satisfy the compiler!
  val contributors: List[Contributor[IdentityState[AbstractAgent]]]
```

I get this error:

```
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:44: overriding value contributors in trait Work of type List[uk.ac.wellcome.models.Contributor[uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]];
[error]  value contributors has incompatible type
[error]   contributors: List[Contributor[MaybeDisplayable[AbstractAgent]]] = Nil,
[error]   ^
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:69: overriding value contributors in trait Work of type List[uk.ac.wellcome.models.Contributor[uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]];
[error]  value contributors has incompatible type
[error]                           contributors: List[Contributor[Displayable[AbstractAgent]]] = Nil,
[error]                           ^
[error] two errors found
[error] (common/compile:compileIncremental) Compilation failed
[error] Total time: 17 s, completed Apr 19, 2018 10:31:25 AM
```

Reply:
you might  want to try making Contributor covariant? `case class Contributor[T <: IdentityState[AbstractAgent]]`

Reply:
Already is, but I think IdentityState is contravariant, so let me try that.

Reply:
So the error above occurs if Contributor is covariant, if it's contravariant (`case class Contributor[T >: IdentityState[AbstractAgent]]`):

```
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:44: overriding value contributors in trait Work of type List[uk.ac.wellcome.models.Contributor[uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]];
[error]  value contributors has incompatible type
[error]   contributors: List[Contributor[MaybeDisplayable[AbstractAgent]]] = Nil,
[error]   ^
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:44: type arguments [uk.ac.wellcome.models.MaybeDisplayable[uk.ac.wellcome.models.AbstractAgent]] do not conform to class Contributor's type parameter bounds [T >: uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]
[error]   contributors: List[Contributor[MaybeDisplayable[AbstractAgent]]] = Nil,
[error]                 ^
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:32: type arguments [uk.ac.wellcome.models.MaybeDisplayable[uk.ac.wellcome.models.AbstractAgent]] do not conform to class Contributor's type parameter bounds [T >: uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]
[error] case class UnidentifiedWork(
[error]            ^
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:69: overriding value contributors in trait Work of type List[uk.ac.wellcome.models.Contributor[uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]];
[error]  value contributors has incompatible type
[error]                           contributors: List[Contributor[Displayable[AbstractAgent]]] = Nil,
[error]                           ^
[error] /repo/common/src/main/scala/uk/ac/wellcome/models/Work.scala:69: type arguments [uk.ac.wellcome.models.Displayable[uk.ac.wellcome.models.AbstractAgent]] do not conform to class Contributor's type parameter bounds [T >: uk.ac.wellcome.models.IdentityState[uk.ac.wellcome.models.AbstractAgent]]
[error]                           contributors: List[Contributor[Displayable[AbstractAgent]]] = Nil,
[error]                                         ^
[error] 5 errors found
[error] (common/compile:compileIncremental) Compilation failed
[error] Total time: 14 s, completed Apr 19, 2018 10:37:06 AM
```

Reply:
`case class Contributor[T >: IdentityState[AbstractAgent]]` doesn't mean contravariant. It mean (I think) T is a superclass of IdentityState[AbstractAgent] rather than the opposite. Sorry I should reread before posting. What I meant in the example above is try this `case class Contributor[+T <: IdentityState[AbstractAgent]]` (note the + before `T`). `IdentityState` is covariant but `Contributor` is not

Reply:
Ah, gotcha. Fifth time lucky? 😉